students/john_rogers/lesson10/donors_fp.py

```diff
@@ -0,0 +1,141 @@
+"""
+Donor classes for mailroom_fp.py
+Author: JohnR
+Version: 2.0
+Last updated: 3/3/2019
+Notes: introducing map, filter and reduce
+"""
+
+from datetime import date
+
+
+class Donor(object):
+    """
+    Class for single donor methods and attributes
+    """
+    def __init__(self, first, last, donations=None):
+        self.first = first
+        self.last = last
+        if isinstance(donations, float):
+            donations = [donations]
+        self.donations = list(donations)
+
+    def __repr__(self):
+        return f'{self.first} {self.last} is' \
+            f' listed as a donor.'
+
+    @property
+    def full_name(self):
+        return f'{self.first} {self.last}'
+
+    @property
+    def total_donations(self):
+        return round(sum(self.donations), 2)
+
+    @property
+    def number_of_donations(self):
+        return len(self.donations)
+
+    @property
+    def get_donations(self):
+        return self.donations
+
+    @property
+    def avg_donation_amount(self):
+        return round(self.total_donations / self.number_of_donations, 2)
+
+    def amped_donations(self, factor):
+        return list(map(lambda x: x * factor, self.get_donations))
+
+    def filtered_donations(self, min_amount=0, max_amount=999999999):
+        return list(filter(lambda x: min_amount <= x <= max_amount,
+                           self.get_donations))
+
+    def new_donation(self, amount):
+        self.donations.append(amount)
+
+    def form_letter(self):
+        """
+        create a form letter
+        :return: form letter filled in with donor and amount
+        """
+        letter = (
+            f'Hey {self.full_name}, thanks for your donations! '
+            f'As of today, {DonorDataBase.today}, you have donated a total of '
+            f'${self.total_donations}.'
+        )
+
+        return letter
+
+
+class DonorDataBase(object):
+    """
+    Class for managing multiple donors
+    """
+    today = date.today()
+
+    def __init__(self, donors=None):
+        self.donors = []
+        if donors:
+            self.donors = donors
+
+    def add_donor(self, donor):
+        self.donors.append(donor)
+
+    def donor_names(self):
+        return [d.full_name for d in self.donors]
+
+    def total_amount_raised(self):
+        return sum([d.total_donations for d in self.donors])
+
+    def total_number_donations(self):
+        return sum([d.number_of_donations for d in self.donors])
+
+    def thank_all(self):
+        for donor in self.donors:
+            letter = Donor.form_letter(donor)
```

Comment:
incorrect usage of class: `form_letter` should be called on an instance.

`donor` is an instance
`Donor` is a class

so should be : `letter = donor.form_letter()`